onnxruntime/core/providers/qnn/builder/opbuilder/transpose_op_builder.cc

```diff
@@ -0,0 +1,122 @@
+#include "core/providers/qnn/builder/qnn_utils.h"
+#include "core/common/safeint.h"
+
+#include "base_op_builder.h"
+
+namespace onnxruntime {
+namespace qnn {
+
+class TransposeOpBuilder : public BaseOpBuilder {
+ public:
+  TransposeOpBuilder() : BaseOpBuilder("TransposeOpBuilder") {}
+  ORT_DISALLOW_COPY_ASSIGNMENT_AND_MOVE(TransposeOpBuilder);
+
+ protected:
+  Status ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
+                                     const NodeUnit& node_unit,
+                                     std::vector<std::string>&& input_names,
+                                     const logging::Logger& logger,
+                                     bool do_op_validation) const override ORT_MUST_USE_RESULT;
+
+ private:
+  Status ProcessPermAttribute(QnnModelWrapper& qnn_model_wrapper,
+                              const NodeUnit& node_unit,
+                              std::vector<std::string>& param_tensor_names) const;
+};
+
+Status TransposeOpBuilder::ProcessPermAttribute(QnnModelWrapper& qnn_model_wrapper,
+                                                const NodeUnit& node_unit,
+                                                std::vector<std::string>& param_tensor_names) const {
+  auto inputs = node_unit.Inputs();
+  std::vector<uint32_t> input_shape;
+  ORT_RETURN_IF_NOT(qnn_model_wrapper.GetOnnxShape(inputs[0].node_arg, input_shape), "Cannot get shape");
+  // set default perm
+  uint32_t rank = static_cast<uint32_t>(input_shape.size());
+  std::vector<int64_t> transpose_perm(rank);
+  for (uint32_t i = 0; i < rank; ++i) {
+    transpose_perm[i] = rank - 1 - i;
+  }
+
+  NodeAttrHelper node_helper(node_unit);
+  transpose_perm = node_helper.Get("perm", transpose_perm);
+  auto perm_size = static_cast<uint32_t>(transpose_perm.size());
+  std::vector<uint32_t> perm_shape{perm_size};
+  std::vector<uint32_t> perm_data;
+  perm_data.resize(perm_size);
+  std::transform(transpose_perm.begin(), transpose_perm.end(), perm_data.begin(),
+                 [](int64_t item) { return SafeInt<uint32_t>(item); });
+
+  QnnParamWrapper transpose_param(node_unit.Index(), node_unit.Name(), QNN_OP_TRANSPOSE_PARAM_PERM,
+                                  std::move(perm_shape), std::move(perm_data));
+  param_tensor_names.push_back(transpose_param.GetParamTensorName());
+  qnn_model_wrapper.AddParamWrapper(std::move(transpose_param));
+
+  return Status::OK();
+}
+
+Status TransposeOpBuilder::ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_wrapper,
+                                                       const NodeUnit& node_unit,
+                                                       std::vector<std::string>&& input_names,
+                                                       const logging::Logger& logger,
+                                                       bool do_op_validation) const {
+  ORT_UNUSED_PARAMETER(logger);
+
+  if (input_names.size() < 1) {
```

Comment:
Is this check necessary for Transpose ops?

Reply:
In case something wrong during ProcessInputs().

---
In reply to: [1304583616](https://github.com/microsoft/onnxruntime/pull/17194#discussion_r1304583616) [](ancestors = 1304583616)